README.md
This Python application is designed to analyze and visualize audio files. It provides a user-friendly interface using Tkinter and incorporates various libraries for audio processing, plotting, and analysis. The application supports loading WAV files, processing them, and displaying different types of plots to aid in understanding the characteristics of the audio.

Features
Load and Process Audio Files

Click the "Load Audio" button to open a file dialog and select an audio file (supported formats: WAV, MP3, AAC).
The selected file will be displayed, indicating a successful load.

Audio Processing:
The application processes the audio file, ensuring it is in WAV format. If not, it converts the file to WAV using the Pydub library.
Information about the resonance frequency, RT60 times, and the duration of the file are then displayed in the GUI.

Plot Audio:

Click the "Plot" button to generate informative plots and display relevant information about the audio file.
The application calculates and prints the total time of the audio, the frequency of the greatest amplitude, and the RT60 (reverberation time) for low, mid, and high frequencies.
Users can cycle through different plots, including Spectrogram, Low Frequency, Mid Frequency, High Frequency, Combined Frequency, and Waveform.

Dependencies
The following Python libraries are required to run the application:

numpy
matplotlib
scipy
pydub
You can install these dependencies using the following command:
python -m pip install numpy matplotlib scipy pydub
or
python -m pip install -r requirements.txt

How to Run
To execute the Audio Analyzer, run the provided Python script. Ensure that you have the necessary dependencies installed. The application will open a graphical user interface, allowing you to interact with the audio analysis tools.

python -m audio_analyzer

Notes
The application uses a GIF image ("music.gif") for visual elements. Ensure the image file is present in the same directory as the script and named properly.

Credits:
 - The music GIF:
   https://wifflegif.com/gifs/706374-music-notes-pixel-art-gif

 - Stackoverflow posts:
  https://stackoverflow.com/questions/23154400/read-the-data-of-a-single-channel-from-a-stereo-wave-file-in-python
  https://stackoverflow.com/questions/28518072/play-animations-in-gif-with-tkinter
  

Changelog.md
0.0.1
Inital project commit

0.1
Add the inital plotting and GUI code

0.2
Added support for multiple file types

0.3
Added waveform plotting
Added RT60 calculation

0.4
Created RT60 plots
Calculated the resonance frequency
Calculated RT60 difference

0.5
Fixed RT60 plots
Added spectrogram plot

0.6
Improved GUI look

0.7
Added a button to cycle the plots

0.8
Added a button for the spectrogram

0.9
Added combined plots

1.0
Added README, requirements.txt, and pyproject.toml
Fixed plots for release

1.1
Added Changelog
Added music GIF

test.py
from colorama import Fore, Back, Style

print("")
print("")
print(f'{Fore.WHITE}$ git log')
print(f'{Fore.YELLOW}commit 83b967c04c445b5cea11bc0aed78acce9987553b ({Fore.CYAN}{Style.BRIGHT}HEAD -> {Fore.GREEN}master{Fore.YELLOW}{Style.NORMAL}, {Fore.GREEN}{Style.BRIGHT}test{Fore.YELLOW}{Style.NORMAL})')
print(f'{Fore.WHITE}Author: Jeremy Goff <jeremygoff03@gmail.com>')
print(f'{Fore.WHITE}Date:   Tus Nov 28 19:12:38 2023 -0500')
print("")
print(f'{Fore.WHITE}     Added data importing')
print("")
print(f'{Fore.YELLOW}commit 2c042d4689cc06f798041b888b0819fb21d41938 ({Fore.CYAN}{Style.BRIGHT}HEAD -> {Fore.GREEN}master{Fore.YELLOW}{Style.NORMAL}, {Fore.GREEN}{Style.BRIGHT}test{Fore.YELLOW}{Style.NORMAL})')
print(f'{Fore.WHITE}Author: Jeremy Goff <jeremygoff03@gmail.com>')
print(f'{Fore.WHITE}Date:   Tus Nov 28 21:31:14 2023 -0500')
print("")
print(f'{Fore.WHITE}     Added data cleaning')
print("")
print(f'{Fore.YELLOW}commit 9cffe506ad550fdab0952ba6adf84466fd853738 ({Fore.CYAN}{Style.BRIGHT}HEAD -> {Fore.GREEN}master{Fore.YELLOW}{Style.NORMAL}, {Fore.GREEN}{Style.BRIGHT}test{Fore.YELLOW}{Style.NORMAL})')
print(f'{Fore.WHITE}Author: Jeremy Goff <jeremygoff03@gmail.com>')
print(f'{Fore.WHITE}Date:   Thu Nov 30 17:06:52 2023 -0500')
print("")
print(f'{Fore.WHITE}     Fixed formatting issues and updated data interface')
print("")

print("")
print("")
print(f'{Fore.WHITE}$ git log')
print(f'{Fore.YELLOW}commit 1de8c72ccc29328abe817e8372c9213b49459094 ({Fore.CYAN}{Style.BRIGHT}HEAD -> {Fore.GREEN}test{Fore.YELLOW}{Style.NORMAL})')
print(f'{Fore.WHITE}Author: Jeremy Goff <jeremygoff03@gmail.com>')
print(f'{Fore.WHITE}Date:   Fri Dec 1 22:43:17 2023 -0500')
print("")
print(f'{Fore.WHITE}     Added visualzation for model and data')
print("")

m.py
import matplotlib
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile

sample_rate, data = wavfile.read("16bit1chan.wav")
spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate,
                                      NFFT=1024, cmap=plt.get_cmap('autumn_r'))
# prints var outputs
# def debugg(fstring):
#     print(fstring) # comment out for prod

def find_target_frequency(freqs, target):
    for x in freqs:
        if x > target:
            break
    return x

def frequency_check():
    # you can choose a frequency which you want to check
    target_frequency = find_target_frequency(freqs, 250)
    index_of_frequency = np.where(freqs == target_frequency)[0][0]
    # find a sound data for a particular frequency
    data_for_frequency = spectrum[index_of_frequency]
    # change a digital signal for a values in decibels
    data_in_db_fun = 10 * np.log10(data_for_frequency)
    return data_in_db_fun

data_in_db = frequency_check()

# plot reverb time on grid
plt.figure()
plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#004bc6')
plt.xlabel('Time (s)')
plt.ylabel('Power (dB)')

# find a index of a max value
index_of_max = np.argmax(data_in_db)
value_of_max = data_in_db[index_of_max]
plt.plot(t[index_of_max], data_in_db[index_of_max], 'go')

# slice array from a max value
sliced_array = data_in_db[index_of_max:]

# find a nearest value
def find_nearest_value(array, value):
    array = np.asarray(array)
    idx = (np.abs(array - value)).argmin()
    return array[idx]

value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max - 5)
index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)
plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')

# slice array from a max -5db
value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max - 25)
index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

plt.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')

rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]
rt60 = 3 * rt20

plt.grid()
plt.show()
print(f'The RT60 reverb time is {round(abs(rt60), 2)} seconds')

r-o.py
import os
import tkinter as tk
from tkinter import filedialog
import librosa
import matplotlib.pyplot as plt
import numpy as np
from pydub import AudioSegment

class AudioAnalyzerApp:
    def __init__(self, root):
        self.root = root
        self.root.title("Audio Analyzer")

        self.file_label = tk.Label(root, text="No file selected")
        self.file_label.pack()

        self.load_button = tk.Button(root, text="Load File", command=self.load_file)
        self.load_button.pack()

        # self.process_button = tk.Button(root, text="Process File", command=self.process_file)
        # self.process_button.pack()

        # self.freq_button = tk.Button(root, text="Alternate Frequency Plot", command=self.toggle_freq_plot)
        # self.freq_button.pack()
        #
        # self.combine_button = tk.Button(root, text="Combine Plots", command=self.combine_plots)
        # self.combine_button.pack()

        self.status_label = tk.Label(root, text="")
        self.status_label.pack()

        self.filename = None
        self.audio_data = None
        self.sample_rate = 10000

        self.waveform_plot = None
        self.freq_plot = None
        self.spec_plot = None
        self.combined_plot = None

        self.current_freq_plot = 0
        # self.file_processed = False

    def load_file(self):
        self.filename = filedialog.askopenfilename(filetypes=[("Audio files", "*.wav;*.mp3;*.aac")])
        self.file_label.config(text=f"File: {os.path.basename(self.filename)}")

        self.process_file()
        # self.toggle_freq_plot()

    def process_file(self):
        if self.filename:
            # Check if the file is a .wav file, if not, convert it
            if not self.filename.lower().endswith(".wav"):
                converted_filename = self.convert_to_wav()
                self.filename = converted_filename

            # Load the audio file using librosa
            self.audio_data, self.sample_rate = librosa.load(self.filename, mono=True, sr=None)

            # Check for meta tags and remove them
            # (Note: This is a placeholder, you may need a more specific implementation based on your requirements)
            self.audio_data = self.remove_meta_tags(self.audio_data)

            # Normalize audio data to the range [-1, 1]
            self.audio_data /= np.max(np.abs(self.audio_data))

            # Display the title value in seconds
            duration = librosa.get_duration(y=self.audio_data, sr=self.sample_rate)
            self.status_label.config(text=f"Duration: {duration:.2f} seconds")

            # Display the waveform
            # self.display_waveform(self.audio_data, self.sample_rate)

            # Compute and display the resonance frequencies
            resonance_freq = self.compute_resonance_frequency(self.audio_data, self.sample_rate)
            self.status_label.config(text=f"Resonance Frequency: {resonance_freq:.2f} Hz")

            # Compute and display RT60 for low, mid, and high frequencies
            _, rt60_low, rt60_mid, rt60_high = self.compute_rt60(self.audio_data, self.sample_rate)
            self.status_label.config(
                text=f"RT60 Low: {rt60_low:.2f} s, RT60 Mid: {rt60_mid:.2f} s, RT60 High: {rt60_high:.2f} s")

            # Display six plots
            self.display_six_plots()

            self.freq_plot = None  # Reset freq_plot

    def convert_to_wav(self):
        # Convert to .wav format using pydub
        audio = AudioSegment.from_file(self.filename)
        converted_filename = self.filename.replace(os.path.splitext(self.filename)[1], ".wav")
        audio.export(converted_filename, format="wav")
        return converted_filename

    def remove_meta_tags(self, audio_data):
        # Placeholder for removing meta tags, modify as needed
        return audio_data

    # def display_six_plots(self):
    #     # Plot waveform
    #     self.display_waveform(self.audio_data, self.sample_rate)
    #
    #     # Plot RT60 for low, mid, and high frequencies
    #     _, rt60_low, rt60_mid, rt60_high = self.compute_rt60(self.audio_data, self.sample_rate)
    #     self.plot_rt60(rt60_low, "Low Frequency RT60", 3)
    #     self.plot_rt60(rt60_mid, "Mid Frequency RT60", 4)
    #     self.plot_rt60(rt60_high, "High Frequency RT60", 5)
    #
    #     # Plot an additional waveform, you can choose another plot type if desired
    #     self.display_waveform(self.audio_data, self.sample_rate)
    def display_six_plots(self):
        # Create a new figure
        plt.figure()

        # Plot waveform
        self.display_waveform(self.audio_data, self.sample_rate, subplot_index=1)

        # Plot RT60 for low, mid, and high frequencies
        # _, rt60_low, rt60_mid, rt60_high = self.compute_rt60(self.audio_data, self.sample_rate)
        # self.plot_rt60(rt60_low, "Low Frequency RT60", 3)
        # self.plot_rt60(rt60_mid, "Mid Frequency RT60", 4)
        # self.plot_rt60(rt60_high, "High Frequency RT60", 5)
        self.display_frequency_content(self.audio_data, self.sample_rate, low_band=(20, 200), subplot_index=2,
                                       label='Low Frequency Content')
        self.display_frequency_content(self.audio_data, self.sample_rate, mid_band=(200, 2000), subplot_index=3,
                                       label='Mid Frequency Content')
        self.display_frequency_content(self.audio_data, self.sample_rate, high_band=(2000, 8000), subplot_index=4,
                                       label='High Frequency Content')

        # Plot an additional waveform, you can choose another plot type if desired
        # self.display_waveform(self.audio_data, self.sample_rate, subplot_index=6)

        # Show the plots
        # plt.tight_layout()
        plt.show()

    def display_frequency_content(self, audio_data, sample_rate, low_band=None, mid_band=None, high_band=None,
                                  subplot_index=None, label=None):
        Zxx = np.abs(librosa.stft(audio_data, hop_length=512, n_fft=self.sample_rate))

        if low_band:
            low_energy = np.sum(np.abs(Zxx[low_band[0]:low_band[1], :]), axis=0)
            plt.subplot(3, 3, subplot_index)
            self.plot_frequency_content(low_energy, sample_rate, label)
            plt.title(f'Frequency Content - {label}')

        if mid_band:
            mid_energy = np.sum(np.abs(Zxx[mid_band[0]:mid_band[1], :]), axis=0)
            plt.subplot(3, 3, subplot_index + 1)
            self.plot_frequency_content(mid_energy, sample_rate, label)
            plt.title(f'Frequency Content - {label}')

        if high_band:
            high_energy = np.sum(np.abs(Zxx[high_band[0]:high_band[1], :]), axis=0)
            plt.subplot(3, 3, subplot_index + 2)
            self.plot_frequency_content(high_energy, sample_rate, label)
            plt.title(f'Frequency Content - {label}')

    def plot_frequency_content(self, energy, sample_rate, label):
        frequency_axis = librosa.fft_frequencies(sr=sample_rate, n_fft=self.sample_rate)
        energy = np.pad(energy, (0, len(frequency_axis) - len(energy)), mode='constant', constant_values=0)
        plt.plot(frequency_axis, energy, label=label)
        plt.xlabel('Frequency (Hz)')
        plt.ylabel('Energy')
        plt.legend()

    def plot_rt60(self, rt60_values, title, subplot_index):
        plt.subplot(3, 2, subplot_index)
        plt.plot(rt60_values)
        plt.title(title)
        plt.xlabel("Frequency Band")
        plt.ylabel("RT60 (s)")
        plt.tight_layout()

    # def display_waveform(self, audio_data, sample_rate):
    #     time = np.arange(0, len(audio_data)) / sample_rate
    #     if self.waveform_plot:
    #         self.waveform_plot.clear()
    #     else:
    #         plt.figure()
    #         self.waveform_plot = plt.subplot(111)
    #     self.waveform_plot.plot(time, audio_data)
    #     self.waveform_plot.set_xlabel('Time (s)')
    #     self.waveform_plot.set_ylabel('Amplitude')
    #     self.waveform_plot.set_title('Waveform')
    #     plt.show()
    def display_waveform(self, audio_data, sample_rate, subplot_index=1):
        time = np.arange(0, len(audio_data)) / sample_rate
        plt.subplot(3, 2, subplot_index)  # Create a subplot within the existing figure
        plt.plot(time, audio_data)
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
        plt.title('Waveform')
        plt.tight_layout()

    def compute_resonance_frequency(self, audio_data, sample_rate):
        # Compute FFT
        fft_result = np.fft.fft(audio_data)

        # Get the corresponding frequencies
        frequencies = np.fft.fftfreq(len(fft_result), 1 / sample_rate)

        # Find the index of the maximum amplitude
        peak_index = np.argmax(np.abs(fft_result))

        # Convert index to frequency
        resonance_freq = np.abs(frequencies[peak_index])

        return resonance_freq

    def compute_rt60(self, audio_data, sample_rate):
        # Compute the short-time Fourier transform (STFT)
        Zxx = np.abs(librosa.stft(audio_data, hop_length=512, n_fft=self.sample_rate))

        # Define frequency bands (you may adjust these as needed)
        low_band = (20, 200)  # Low frequencies
        mid_band = (200, 2000)  # Mid frequencies
        high_band = (2000, 8000)  # High frequencies

        # Compute the energy in each frequency band
        frames_2s = librosa.time_to_frames(2, sr=sample_rate, hop_length=512)
        frames_4s = librosa.time_to_frames(4, sr=sample_rate, hop_length=512)

        energy_low = np.sum(np.abs(Zxx[low_band[0]:low_band[1], frames_2s:frames_4s]))
        energy_mid = np.sum(np.abs(Zxx[mid_band[0]:mid_band[1], frames_2s:frames_4s]))
        energy_high = np.sum(np.abs(Zxx[high_band[0]:high_band[1], frames_2s:frames_4s]))

        # Compute RT60 using the energy in each band
        rt60_low = -60 / np.mean(20 * np.log10(energy_low / np.max(energy_low)))
        rt60_mid = -60 / np.mean(20 * np.log10(energy_mid / np.max(energy_mid)))
        rt60_high = -60 / np.mean(20 * np.log10(energy_high / np.max(energy_high)))

        return Zxx, rt60_low, rt60_mid, rt60_high

    # def toggle_freq_plot(self):
    #     self.current_freq_plot = (self.current_freq_plot + 1) % 3
    #     self.display_freq_plot()
    #     # if self.file_processed:
    #     #     self.current_freq_plot = (self.current_freq_plot + 1) % 3
    #     #     self.display_freq_plot()
    #     # else:
    #     #     self.status_label.config(text="Please process the file first.")

    # def display_freq_plot(self):
    #     # if self.file_processed:
    #     freq_range = ["Low", "Mid", "High"]
    #     freq_label = freq_range[self.current_freq_plot]
    #
    #     # Compute STFT using librosa
    #     Zxx, rt60_low, rt60_mid, rt60_high = self.compute_rt60(self.audio_data, len(self.audio_data))
    #
    #     make_new_plot = False
    #     if self.freq_plot:
    #         self.freq_plot.clear()
    #     else:
    #         make_new_plot = True
    #         plt.figure()
    #         self.freq_plot = plt.subplot(211)
    #
    #         # Display spectrogram
    #         librosa.display.specshow(librosa.amplitude_to_db(Zxx, ref=np.max), y_axis='log', x_axis='time')
    #         plt.colorbar(format='%+2.0f dB')
    #         plt.title('Spectrogram of Waveform')
    #
    #     # Display frequency plot
    #     if make_new_plot:
    #         self.freq_plot = plt.subplot(212)
    #         freq_data = np.mean(Zxx, axis=1)  # Extract mean amplitude along the time axis
    #         freq_axis = librosa.fft_frequencies(sr=len(self.audio_data), n_fft=1024)
    #         plt.plot(freq_axis, freq_data, label=f'{freq_label} Frequency Plot')
    #         plt.xlabel('Frequency (Hz)')
    #         plt.ylabel('Amplitude')
    #         plt.title(f'{freq_label} Frequency Plot')
    #
    #     # Display RT60 values
    #     # plt.axvline(rt60_low, color='r', linestyle='--', label=f'RT60 Low: {rt60_low:.2f} s')
    #     # plt.axvline(rt60_mid, color='g', linestyle='--', label=f'RT60 Mid: {rt60_mid:.2f} s')
    #     # plt.axvline(rt60_high, color='b', linestyle='--', label=f'RT60 High: {rt60_high:.2f} s')
    #
    #     plt.legend()
    #     plt.tight_layout()
    #     plt.show()
    #     # else:
    #     #     self.status_label.config(text="Please load and process the file first.")
    #
    # def combine_plots(self):
    #     if self.waveform_plot and self.freq_plot:
    #         # Placeholder for combining plots
    #         # You should replace this with your actual combination code
    #         if self.combined_plot:
    #             self.combined_plot.clear()
    #         else:
    #             plt.figure()
    #             self.combined_plot = plt.subplot(111)
    #
    #         # Plot waveform
    #         self.combined_plot.plot(self.waveform_plot.get_lines()[0].get_xdata(),
    #                                 self.waveform_plot.get_lines()[0].get_ydata(), label='Waveform')
    #
    #         # Plot frequency data
    #         freq_data = np.random.rand(100)  # Replace with your frequency data
    #         freq_axis = np.linspace(0, 1, len(freq_data))
    #         self.combined_plot.plot(freq_axis, freq_data, label='Frequency Plot')
    #
    #         self.combined_plot.set_xlabel('Time/Frequency')
    #         self.combined_plot.set_ylabel('Amplitude')
    #         self.combined_plot.set_title('Combined Plot')
    #         self.combined_plot.legend()
    #         plt.show()
    #     else:
    #         self.status_label.config(text="Please process the file first.")

    # def plot_waveform(self):
    #     plt.figure(figsize=(10, 4))
    #     time = np.arange(0, len(self.audio_data)) / self.sample_rate
    #     plt.plot(time, self.audio_data)
    #     plt.title('Waveform')
    #     plt.xlabel('Time (s)')
    #     plt.ylabel('Amplitude')
    #     plt.tight_layout()
    #     plt.savefig("waveform.png")
    #
    #     # Display the waveform plot
    #     self.plot_image("waveform.png")
    #
    # def compute_resonance_frequency(self):
    #     # Your resonance frequency computation logic here
    #     # This is just a placeholder, replace with your actual implementation
    #     return np.random.uniform(50, 1000)
    #
    # def plot_image(self, image_path):
    #     img = tk.PhotoImage(file=image_path)
    #     panel = tk.Label(self.plot_frame, image=img)
    #     panel.image = img
    #     panel.pack()
    #
    # def plot_decibel_vs_frequency(self):
    #     plt.figure(figsize=(10, 4))
    #     D = librosa.amplitude_to_db(np.abs(librosa.stft(self.audio_data)), ref=np.max)
    #     librosa.display.specshow(D, sr=44100, x_axis='time', y_axis='log')
    #     plt.colorbar(format='%+2.0f dB')
    #     plt.title('Spectrogram')
    #     plt.xlabel('Time (s)')
    #     plt.ylabel('Frequency (Hz)')
    #     plt.tight_layout()
    #     plt.savefig("decibel_vs_frequency.png")
    #
    #     # Display the decibel vs frequency plot
    #     self.plot_image("decibel_vs_frequency.png")
    #
    # def plot_line_chart(self):
    #     # Compute the frequency content and mean decibels
    #     frequencies, mean_decibels = self.compute_frequency_content()
    #
    #     # Plot Line Chart
    #     plt.figure(figsize=(10, 4))
    #     plt.plot(frequencies, mean_decibels)
    #     plt.title('Decibels vs Frequency')
    #     plt.xlabel('Frequency (Hz)')
    #     plt.ylabel('Mean Decibels')
    #     plt.tight_layout()
    #     plt.savefig("line_chart.png")
    #
    #     # Display the line chart plot
    #     self.plot_image("line_chart.png")
    #
    # def compute_frequency_content(self):
    #     # Compute the frequency content and mean decibels
    #     D = librosa.amplitude_to_db(np.abs(librosa.stft(self.audio_data)), ref=np.max)
    #     frequencies = librosa.core.fft_frequencies(sr=44100)
    #     mean_decibels = np.mean(D, axis=1)
    #     return frequencies, mean_decibels
    #
    # def plot_frequency_ranges(self):
    #     # Compute the frequency content and mean decibels
    #     frequencies, mean_decibels = self.compute_frequency_content()
    #
    #     # Define frequency ranges
    #     low_freq_range = (0, 1000)
    #     mid_freq_range = (1000, 5000)
    #     high_freq_range = (5000, 20000)
    #
    #     # Filter frequencies and decibels for each range
    #     low_freq_indices = np.where((frequencies >= low_freq_range[0]) & (frequencies < low_freq_range[1]))[0]
    #     mid_freq_indices = np.where((frequencies >= mid_freq_range[0]) & (frequencies < mid_freq_range[1]))[0]
    #     high_freq_indices = np.where((frequencies >= high_freq_range[0]) & (frequencies < high_freq_range[1]))[0]
    #
    #     low_freq_values = mean_decibels[low_freq_indices]
    #     mid_freq_values = mean_decibels[mid_freq_indices]
    #     high_freq_values = mean_decibels[high_freq_indices]
    #
    #     # Plot Line Chart for Frequency Ranges
    #     plt.figure(figsize=(10, 4))
    #
    #     plt.plot(frequencies[low_freq_indices], low_freq_values, label='Low Frequencies')
    #     plt.plot(frequencies[mid_freq_indices], mid_freq_values, label='Mid Frequencies')
    #     plt.plot(frequencies[high_freq_indices], high_freq_values, label='High Frequencies')
    #
    #     plt.title('Decibels vs Frequency Ranges')
    #     plt.xlabel('Frequency (Hz)')
    #     plt.ylabel('Mean Decibels')
    #     plt.legend()
    #     plt.tight_layout()
    #     plt.savefig("frequency_ranges_chart.png")
    #
    #     # Display the line chart for frequency ranges
    #     self.plot_image("frequency_ranges_chart.png")

if __name__ == "__main__":
    root = tk.Tk()
    app = AudioAnalyzerApp(root)
    root.mainloop()

r.py
import os
import tkinter as tk
from tkinter import *
from tkinter import filedialog
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile
from pydub import AudioSegment

class AudioAnalyzerApp:
    def __init__(self, root):
        self.root = root
        self.root.title("Audio Analyzer")

        self.gif_label = Label(root)
        self.frame_count = 9
        self.gif_frames = [PhotoImage(file='music.gif', format='gif -index %i' % (i)) for i in range(self.frame_count)]

        self.file_path = None
        self.audio_data = None
        self.freq_plot = None

        self.plot_count = 6
        self.current_freq_plot = 0
        self.frequency_values = [250, 1000, 8000]

        self.create_widgets()

    def create_widgets(self):
        self.gif_label.pack()
        # Load File Button
        self.current_plot_label = tk.Label(self.root, text="Current Plot: None")
        self.current_plot_label.pack(pady=10)

        self.file_label = tk.Label(self.root, text="Current File: None")
        self.file_label.pack()

        load_button = tk.Button(self.root, text="Load File", command=self.load_file)
        load_button.pack()

        self.toggle_freq_button = tk.Button(root, text="Change Plot", command=self.show_next_plot)
        self.toggle_freq_button.pack(pady=10)

        self.show_spectrogram_button = tk.Button(root, text="Show spectrogram", command=self.show_spectrogram)
        self.show_spectrogram_button.pack()

        # Display Length
        self.length_label = tk.Label(self.root, text="Length: None seconds")
        self.length_label.pack()

        # Display Resonance Frequency
        self.resonance_label = tk.Label(self.root, text="Resonance: None Hz")
        self.resonance_label.pack()

        # Display Low, Mid, High Frequencies
        self.frequency_labels = tk.Label(self.root, text="Low: None | Mid: None | High: None")
        self.frequency_labels.pack()

        # Waveform Plot
        self.plot_frame = tk.Frame(self.root)
        self.plot_frame.pack()

    def load_file(self):
        file_path = filedialog.askopenfilename(filetypes=[("Audio Files", "*.wav;*.mp3;*.aac")])

        # if the file is a valid process the file
        if file_path:
            self.file_path = file_path
            self.file_label.config(text=f"File: {os.path.basename(self.file_path)}")

            self.process_audio()

    def process_audio(self):
        # Check if the file is a .wav file
        if not self.file_path.lower().endswith('.wav'):
            wav_path = self.convert_to_wav()
        else:
            wav_path = self.file_path

        # Read .wav file
        self.sample_rate, audio_data = wavfile.read(wav_path)

        # check for more than one channel and remove them
        if audio_data[1].size > 1:
            audio_data = audio_data[:, 0]

        # normalize audio data so it can be better shown on a plot
        self.audio_data = audio_data / np.max(np.abs(audio_data), axis=0)
        duration = len(audio_data) / self.sample_rate

        # Display length
        self.length_label.config(text=f"Length: {duration:.2f} seconds")

        # Display plot
        self.display_plot()

    def display_plot(self):
        plot_name = self.get_plot_name(self.current_freq_plot)
        self.current_plot_label.config(text=f"Current Plot: {plot_name}")
        if self.current_freq_plot == 0:
            # selected plot is spectrogram
            self.plot_spectrogram()
            return
        if self.current_freq_plot == 1 or self.current_freq_plot == 2 or self.current_freq_plot == 3:
            # selected plot is low, mid, or high
            target = self.frequency_values[self.current_freq_plot - 1]
            self.update_figure_plot()
            self.plot_frequency(self.frequency_check(target), self.get_plot_color(self.current_freq_plot), plot_name)
            plt.show()
            return
        if self.current_freq_plot == 4:
            # selected plot is combined
            self.plot_combined()
            return
        if self.current_freq_plot == 5:
            # selected plot is waveform
            self.plot_waveform()
            return

    def get_plot_name(self, value):
        if value == 0: return "Spectrogram"
        if value == 1: return "Low"
        if value == 2: return "Mid"
        if value == 3: return "High"
        if value == 4: return "Combined"
        if value == 5: return "Waveform"

    def get_plot_color(self, value):
        if value == 0: return ""
        if value == 1: return "r"
        if value == 2: return "g"
        if value == 3: return "b"
        if value == 4: return ""
        if value == 5: return "b"

    def show_next_plot(self):
        if not self.file_path:
            return
        self.current_freq_plot = (self.current_freq_plot + 1) % self.plot_count
        self.display_plot()

    def show_spectrogram(self):
        if not self.file_path:
            return
        self.current_freq_plot = 0
        self.display_plot()

    def convert_to_wav(self):
        # Convert to .wav using pydub
        audio = AudioSegment.from_file(self.file_path)
        wav_path = os.path.splitext(self.file_path)[0] + ".wav"
        audio.export(wav_path, format="wav")
        return wav_path

    def find_target_frequency(self, freqs, target):
        for x in freqs:
            if x > target:
                break
        return x

    def frequency_check(self, target):
        # you can choose a frequency which you want to check
        target_frequency = self.find_target_frequency(self.frequencies, target)
        index_of_frequency = np.where(self.frequencies == target_frequency)[0][0]
        # find a sound data for a particular frequency
        data_for_frequency = self.spectrum[index_of_frequency]
        # change a digital signal for a values in decibels
        data_in_db_fun = 10 * np.log10(data_for_frequency)
        return data_in_db_fun

    def find_nearest_value(self, array, value):
        array = np.asarray(array)
        idx = (np.abs(array - value)).argmin()
        return array[idx]

    def update_figure_plot(self):
        # reuse the already shown plot or make a new one
        if self.freq_plot:
            self.freq_plot.clear()
        else:
            plt.figure()
            self.freq_plot = plt.subplot(111)

    def plot_spectrogram(self):
        self.update_figure_plot()

        self.spectrum, self.frequencies, self.time_values, _ = plt.specgram(self.audio_data, Fs=self.sample_rate,
                                              NFFT=1024, cmap=plt.get_cmap('autumn_r'))

        plt.show()

    def plot_frequency(self, db_data, color, label):
        plt.plot(self.time_values, db_data, linewidth=1, alpha=0.7, color=f'{color}', label=f'{label}')
        plt.xlabel('Time (s)')
        plt.ylabel('Power (dB)')

        index_of_max = np.argmax(db_data)
        self.resonance_label.config(text=f"Resonance: {np.round(self.frequencies[index_of_max], 3)} Hz")

        value_of_max = db_data[index_of_max]
        plt.plot(self.time_values[index_of_max], value_of_max, 'go')

        sliced_array = db_data[index_of_max:]

        value_of_max_less_5 = self.find_nearest_value(sliced_array, value_of_max - 5)
        index_of_max_less_5 = np.where(db_data == value_of_max_less_5)
        plt.plot(self.time_values[index_of_max_less_5], db_data[index_of_max_less_5], 'yo')

        value_of_max_less_25 = self.find_nearest_value(sliced_array, value_of_max - 25)
        index_of_max_less_25 = np.where(db_data == value_of_max_less_25)

        plt.plot(self.time_values[index_of_max_less_25], db_data[index_of_max_less_25], 'ro')

        time_average = (self.time_values[index_of_max] + self.time_values[index_of_max_less_5] + self.time_values[index_of_max_less_25]) / 3
        rt20 = (self.time_values[index_of_max_less_5] - self.time_values[index_of_max_less_25])[0]
        rt60 = 3 * rt20

        self.frequency_labels.config(text=f"Average: {np.round(time_average[0], 3)}, Diff: {np.round(time_average[0] - 0.5, 3)}")

    def plot_combined(self):
        self.update_figure_plot()

        data = [self.frequency_check(self.frequency_values[0]), self.frequency_check(self.frequency_values[1]), self.frequency_check(self.frequency_values[2])]

        self.plot_frequency(data[0], self.get_plot_color(1), self.get_plot_name(1))
        self.plot_frequency(data[1], self.get_plot_color(2), self.get_plot_name(2))
        self.plot_frequency(data[2], self.get_plot_color(3), self.get_plot_name(3))
        plt.show()

    def plot_waveform(self):
        self.update_figure_plot()

        time = np.arange(0, len(self.audio_data)) / self.sample_rate
        plt.plot(time, self.audio_data)
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
        plt.title('Waveform')
        plt.tight_layout()
        plt.show()

    def update_gif(self, ind):
        frame = self.gif_frames[ind]
        ind += 1
        if ind == self.frame_count:
            ind = 0
        self.gif_label.configure(image=frame)
        root.after(100, self.update_gif, ind)

if __name__ == "__main__":
    root = tk.Tk()
    app = AudioAnalyzerApp(root)
    root.after(0, app.update_gif, 0)
    root.mainloop()